File: src/credentials.ts
import { execFileSync } from 'node:child_process';

/**
 * An op:// reference to a 1Password item field.
 * Format: op://<vault>/<item>/<field>
 */
export interface OpReference {
  ref: string;
}

export function isOpReference(value: string): boolean {
  return value.startsWith('op://');
}

/**
 * Credential scope declared in principles.md.
 * Maps an action to the op:// references it requires.
 */
export interface CredentialScope {
  action: string;
  refs: string[];
}

/**
 * Resolve an op:// reference via the 1Password CLI.
 * Returns the resolved value. Throws if the vault is locked or the reference is invalid.
 *
 * IMPORTANT: The resolved value must NEVER be logged, written to disk, or included
 * in journal entries. It is used only at the moment of dispatch.
 */
export function resolveOpReference(ref: string): string {
  if (!isOpReference(ref)) {
    throw new Error(`not an op:// reference: ${ref}`);
  }
  try {
    const result = execFileSync('op', ['read', ref], {
      encoding: 'utf8',
      timeout: 10_000,
      stdio: ['pipe', 'pipe', 'pipe'],
    });
    return result.trim();
  } catch (err) {
    const message = (err as Error).message ?? '';
    if (message.includes('not signed in') || message.includes('session expired') || message.includes('locked')) {
      throw new Error(`vault locked or not signed in — cannot resolve ${ref}`);
    }
    throw new Error(`failed to resolve ${ref}: ${message}`);
  }
}

/**
 * Check whether the 1Password vault is currently unlocked by attempting
 * a lightweight operation. Returns true if unlocked, false if locked.
 */
export function isVaultUnlocked(): boolean {
  try {
    execFileSync('op', ['whoami'], {
      encoding: 'utf8',
      timeout: 5_000,
      stdio: ['pipe', 'pipe', 'pipe'],
    });
    return true;
  } catch {
    return false;
  }
}

/**
 * Injectable credential resolver interface for testing.
 * Production uses resolveOpReference; tests inject a fake.
 */
export interface CredentialResolver {
  resolve(ref: string): string;
  isUnlocked(): boolean;
}

/** Production credential resolver backed by 1Password CLI. */
export function createOpResolver(): CredentialResolver {
  return {
    resolve: resolveOpReference,
    isUnlocked: isVaultUnlocked,
  };
}

/**
 * Check that all required credential scopes for an action are resolvable.
 * Returns { allowed: true } or { allowed: false, reason: string }.
 */
export function checkCredentialScopes(
  action: string,
  scopes: CredentialScope[],
  resolver: CredentialResolver,
): { allowed: true } | { allowed: false; reason: string } {
  const scope = scopes.find((s) => s.action === action);
  if (!scope) {
    // No credential scope declared — action doesn't require credentials
    return { allowed: true };
  }

  if (!resolver.isUnlocked()) {
    return { allowed: false, reason: 'vault is locked — cannot resolve credentials' };
  }

  for (const ref of scope.refs) {
    try {
      resolver.resolve(ref);
    } catch (err) {
      return { allowed: false, reason: (err as Error).message };
    }
  }

  return { allowed: true };
}

File: python/steward/credentials.py
"""1Password CLI credential resolver. Executor-side only — never in LLM process."""
from __future__ import annotations

import subprocess
from dataclasses import dataclass
from typing import Protocol

from steward.rules import CredentialScopeDecl


def is_op_reference(value: str) -> bool:
    return value.startswith("op://")


def resolve_op_reference(ref: str) -> str:
    """Resolve via `op read`. Never log the return value."""
    if not is_op_reference(ref):
        raise ValueError(f"not an op:// reference: {ref}")
    try:
        result = subprocess.run(
            ["op", "read", ref],
            capture_output=True,
            text=True,
            timeout=10,
            check=True,
        )
        return result.stdout.strip()
    except subprocess.CalledProcessError as e:
        stderr = e.stderr or ""
        if any(s in stderr for s in ("not signed in", "session expired", "locked")):
            raise RuntimeError(f"vault locked or not signed in — cannot resolve {ref}") from e
        raise RuntimeError(f"failed to resolve {ref}: {stderr.strip()}") from e
    except FileNotFoundError as e:
        raise RuntimeError("op CLI not found on PATH") from e


def is_vault_unlocked() -> bool:
    try:
        subprocess.run(
            ["op", "whoami"],
            capture_output=True,
            text=True,
            timeout=5,
            check=True,
        )
        return True
    except (subprocess.CalledProcessError, FileNotFoundError, subprocess.TimeoutExpired):
        return False


class CredentialResolver(Protocol):
    def resolve(self, ref: str) -> str: ...
    def is_unlocked(self) -> bool: ...


@dataclass
class OpResolver:
    def resolve(self, ref: str) -> str:
        return resolve_op_reference(ref)

    def is_unlocked(self) -> bool:
        return is_vault_unlocked()


@dataclass
class ScopeCheck:
    allowed: bool
    reason: str | None = None


def check_credential_scopes(
    action: str,
    scopes: list[CredentialScopeDecl],
    resolver: CredentialResolver,
) -> ScopeCheck:
    scope = next((s for s in scopes if s.action == action), None)
    if scope is None:
        return ScopeCheck(allowed=True)
    if not resolver.is_unlocked():
        return ScopeCheck(allowed=False, reason="vault is locked — cannot resolve credentials")
    for ref in scope.refs:
        try:
            resolver.resolve(ref)
        except Exception as e:
            return ScopeCheck(allowed=False, reason=str(e))
    return ScopeCheck(allowed=True)
